src/typecheck.rs

```diff
@@ -109,337 +232,234 @@ impl PrimitiveLike for ValueEq {
+                    .collect::<Vec<_>>();
+                atoms.append(&mut equality_checks);
+                Query { atoms }
+            }
+            Fact::Fact(expr) => {
+                let atoms = self.flatten_expr(expr).1;
+                Query { atoms }
             }
         }
+    }
 
-        // replace canonical things in the actions
-        let res_actions = actions.iter().map(|a| a.replace_canon(&canon)).collect();
-        for (var, _expr) in canon {
-            self.types.remove(&var);
+    pub fn lower(&mut self, rule: &Rule) -> Result<UnresolvedCoreRule, Vec<TypeError>> {
+        let facts = &rule.body;
+        let actions = &rule.head;
+        let mut query = Query::default();
+        for fact in facts {
+            query += self.flatten_fact(fact);
         }
+        // let desugar = ;
+        Ok(UnresolvedCoreRule {
+            body: query,
+            head: Actions(flatten_actions(actions, &mut self.egraph.desugar)),
+        })
+    }
 
-        let get_leaf = |id: &Id| -> AtomTerm {
-            let mk = || AtomTerm::Var(Symbol::from(format!("?__{}", id)));
-            match leaves.get(id) {
-                Some(Expr::Var(v)) => {
-                    if let Some((_ty, _value, _ts)) = self.egraph.global_bindings.get(v) {
-                        AtomTerm::Global(*v)
-                    } else {
-                        AtomTerm::Var(*v)
-                    }
-                }
-                Some(Expr::Lit(l)) => AtomTerm::Value(self.egraph.eval_lit(l)),
-                _ => mk(),
-            }
-        };
-
-        let mut query = Query::default();
-        let mut query_eclasses = HashSet::<Id>::default();
-        // Now we can fill in the nodes with the canonical leaves
-        for (node, id) in &self.nodes {
-            match node {
-                ENode::Func(f, ids) => {
-                    let args = ids.iter().chain([id]).map(get_leaf).collect();
-                    for id in ids {
-                        query_eclasses.insert(*id);
-                    }
-                    query.atoms.push(Atom { head: *f, args });
-                }
-                ENode::Prim(p, ids) => {
-                    let mut args = vec![];
-                    for child in ids {
-                        let leaf = get_leaf(child);
-                        if let AtomTerm::Var(v) = leaf {
-                            if self.egraph.global_bindings.contains_key(&v) {
-                                args.push(AtomTerm::Value(self.egraph.global_bindings[&v].1));
-                                continue;
-                            }
+    // TODO: should this be in ResolvedCoreRule
+    pub fn canonicalize(&self, rule: UnresolvedCoreRule) -> UnresolvedCoreRule {
+        let mut result_rule = rule;
+        loop {
+            let mut to_subst = None;
+            for atom in result_rule.body.atoms.iter() {
+                if atom.head == "value-eq".into() && atom.args[0] != atom.args[1] {
```

Comment:
I don't like this special case- instead of using the primitive `value-eq`, equality constraints should be specially represented in the data structure. For that reason, I have equality constraints in the `NormFact` data structure.

Reply:
Still think we shouldn't have `value-eq`, see my comment above.